compiler/qsc_fir/src/ty.rs

```diff
@@ -178,20 +178,77 @@ fn instantiate_arrow_ty<'a>(
     })
 }
 
-impl Display for GenericParam {
+impl Display for TypeParameter {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         match self {
-            GenericParam::Ty => write!(f, "type"),
-            GenericParam::Functor(min) => write!(f, "functor ({min})"),
+            TypeParameter::Ty { name, bounds } => {
+                write!(f, "type ({name}){bounds}")
+            }
+            TypeParameter::Functor(min) => write!(f, "functor ({min})"),
+        }
+    }
+}
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct ClassConstraints(pub Box<[ClassConstraint]>);
+
+impl std::fmt::Display for ClassConstraints {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        if self.0.is_empty() {
+            Ok(())
+        } else {
+            let bounds = self
+                .0
+                .iter()
+                .map(std::string::ToString::to_string)
+                .collect::<Vec<_>>()
+                .join(", ");
+            write!(f, "{bounds}")
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub enum ClassConstraint {
+    #[default]
```

Comment:
Is it correct to say that the default of a `ClassConstraint` is the `Eq` constraint? Where do we use this?

Reply:
This is a bad copy-paste on my part -- we have to have a default in the parser for parser recovery (a pattern I'm not actually very happy with) and I copy/pasted it here.

It hasn't been used anywhere so I removed it. Good catch, thank you.